### PR TITLE
[FIX] website_sale: add product snippet filter on all websites

### PR DESCRIPTION
**Pull request by momegahed**
If applied, this commit will fix the following bug by changing
the implementation of adding the product filter to websites
to include the rest of the websites not only the default one

Steps to reproduce:

1- install website module
2- create more than 1 website
3- install website_sale module
4- the dynamic product snippet only works on the default website

Bug:

the ```website.snippet.filter``` for products is added to the default
website only

Fix:

convert to post_init_hook and include all websites

related to: https://github.com/odoo/odoo/commit/01fafecf98aba5886c5fe836e2a81f2551155f35

DON'T forward port to v15 onwards, the website_id field is not required
 anymore for website.snippet.filter which makes them global by default

OPW-2754340

### CODE REVIEW

**Comment by momegahed** [comment]
@simongoffin

File: addons/website_sale/__manifest__.py [hunk 1]
```diff
@@ -33,6 +33,7 @@
     'qweb': ['static/src/xml/*.xml'],
     'installable': True,
     'application': True,
+    'post_init_hook': 'post_init_hook',
```
Comment (bso-odoo) [inline addons/website_sale/__manifest__.py:36]: I believe `post_init_hook` is only run when a module is first installed. (I might be wrong)
What happens for users already using v14 with the problem ? I think this will not be called when they update (`odoo-bin -u website_sale`) the module. Do we rely on support to tell them to create the missing filter records ?

**Review by bso-odoo — COMMENTED** [review]
Here are a few  small comments in the code.

Also the commit message does not describe the change - and does not start with a verb (https://www.odoo.com/documentation/14.0/developer/misc/other/guidelines.html#commit-message-header).
What about something like "add product snippet filter on all websites" ?

The explanation for not forwarding to v15 is a bit fishy: AFAIK v15 still requires the `filter_id`, but those filters do not require a `website_id` anymore - making them global by default.
FYI - you can actually use commands to instruct robodoo about the forwarding (https://github.com/odoo/odoo/wiki/Mergebot#commands)

File: addons/website_sale/__init__.py [hunk 2]
```diff
@@ -17,3 +17,19 @@ def uninstall_hook(cr, registry):
...
+    for website in env['website'].search([]):
+        action = env.ref('website_sale.dynamic_snippet_products_action', raise_if_not_found=False)
+        if action:
+            env['website.snippet.filter'].create({
+                'action_server_id': action.id,
+                'field_names': 'display_name,description_sale,image_512,list_price',
+                'limit': 16,
+                'name': _('Products'),
+                'website_id': website.id,
+            })
```
[the remark this answers is not in the document]
Comment (momegahed) [reply addons/website_sale/__init__.py:35]: no I didn't but why push it to 15.0 if its irrelevant anyways ?

File: addons/website_sale/__manifest__.py (hunk 1 above)
Comment (momegahed) [reply addons/website_sale/__manifest__.py:36]: yes, note that this problem occurs only in the websites created **before** the installation of website_sale. so for these website either website_sale will have to be reinstalled or the website have to be recreated. if you have a better idea on how we can also include the current faulty website in the fix please let me know

**Comment by momegahed** [comment]
up to 14.0

File: addons/website_sale/__init__.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (bso-odoo) [reply addons/website_sale/__init__.py:35]: This is not what I meant, sorry.

Here is the scenario I had in mind:
- user uses this version => there are new filter records compared to before the fix
- user drops the product snippet(s) into its pages
- user upgrades odoo to v15
=> do the snippets dropped in v14 still work after the upgrade ?

If it is not the case, we might have to write an upgrade script to convert to v14 dropped snippet into one that is compatible with v15.

[the remark this answers is not in the document]
Comment (momegahed) [reply addons/website_sale/__init__.py:35]: well... this fix doesn't change the behavior of product snippet filters. if before the fix the user installed eCommerce first and then created a second website it will have the filter as normal. so I am not introducing new filter records, I am just adding them to all websites created before the installation. 

tldr:
this PR doesn't change the answer to your question. if snippets dropped in v14 work after the upgrade, they will still work after the upgrade, if not, then not.

**Comment by momegahed** [comment]
@fw-bot  up to 14.0

**Review by bso-odoo — APPROVED** [review]
lgtm, thank you :+1:

File: addons/website_sale/data/data.xml [hunk 3]
```diff
@@ -96,14 +96,7 @@ for product in products:
 response = res_products
             </field>
         </record>
-        <!-- Dynamic Filter -->
-        <record id="dynamic_filter_demo_products" model="website.snippet.filter">
```
Comment (rdeodoo) [inline addons/website_sale/data/data.xml:100]: You can't delete records in stable, people might have code that relies on this

Probably add something like this in the post_init

```python
filter = self.env.ref('website_sale.dynamic_filter_demo_products', False)
if action and filter:
        for website in env['website'].search([('id', '!=', filter.website_id.id)]):
```

File: addons/website_sale/__init__.py [hunk 4]
```diff
@@ -17,3 +17,19 @@ def uninstall_hook(cr, registry):
...
+    action = env.ref('website_sale.dynamic_snippet_products_action', raise_if_not_found=False)
+    if action:
+        for website in env['website'].search([]):
+            env['website.snippet.filter'].create({
+                'action_server_id': action.id,
+                'field_names': 'display_name,description_sale,image_512,list_price',
+                'limit': 16,
+                'name': _('Products'),
+                'website_id': website.id,
+            })
```
Comment (rdeodoo) [inline addons/website_sale/__init__.py:35]: Copy pasting code is generally not great. This should probably be encapsulated and reused. It would be a lot easier to figure when someone will read the code, as greping the new method name would return 2 places, the one here and the one in bootstrap_website.

This is especially the case when it is not mentioned, it makes the review harder than it should, here you should mention commit https://github.com/odoo/odoo/commit/01fafecf98aba5886c5fe836e2a81f2551155f35

**Review by rdeodoo — CHANGES_REQUESTED** [review]
While the approach of the current implementation is fine to me, we could also just overide the create method of the `website.snippet.filter` model, and if we are in install mode, we could create that field for all other websites.
That would make the fix generic, including possible custo.
That might be overkill as this only target 14.0 tho, so maybe keep it like this and just process the review I left

File: addons/website_sale/__init__.py (hunk 4 above)
Comment (momegahed) [reply addons/website_sale/__init__.py:35]: well, since most of the code here is just the parameters of the create method, I was thinking it doesn't make much since to encapsulate it as it is somehow already encapsulated in the create method?

All the rest is just logic and parameters, encapsulating the logic is also not useful because it is a bit different from elsewhere (due to searching all websites and checking the action first)

What do you think?
also do you mean mention the commit in a comment in the code or in the commit message?

Comment (rdeodoo) [reply addons/website_sale/__init__.py:35]: I would have still encapsulated, but I got your point.
If I were to keep it like you did, I would at least add a comment here
```py
# Do the same as in _bootstrap_snippet_filters()
```
And same on `_bootstrap_snippet_filters()`, add 
```py
# The same behavior is done in the post_init hook
```

That really is just to ease code reading later, if you agree.

> also do you mean mention the commit in a comment in the code or in the commit message?

Commit msg :)

File: addons/website_sale/__init__.py [hunk 5]
```diff
@@ -17,3 +17,22 @@ def uninstall_hook(cr, registry):
...
+
+    if action and filter:
+        for website in env['website'].search([('id', '!=', filter.website_id.id)]):
+            env['website.snippet.filter'].create({
+                'action_server_id': action.id,
+                'field_names': 'display_name,description_sale,image_512,list_price',
+                'limit': 16,
+                'name': _('Products'),
+                'website_id': website.id,
+            })
```
Comment (rdeodoo) [inline addons/website_sale/__init__.py:38]: This is enough?
```suggestion
    filter = env.ref('website_sale.dynamic_filter_demo_products', False)
    if filter:
        for website in env['website'].search([('id', '!=', filter.website_id.id)]):
            filter.copy({'website_id': website.id})
```

**Review by rdeodoo — APPROVED** [review]
I pushed a new diff, we don't need to create with all properties the record, a copy is enough.
I also changed the `'''` to `"""` as it should be in docstring (I changed the already existing ones so we don't copy paste those again, note that I introduced them myself :grimacing: )
I rewrote the commit message, you can see the change by comparing the PR msg and the commit one.

Do you see any issue with the changes I made?

File: addons/website_sale/__init__.py (hunk 5 above)
Comment (momegahed) [reply addons/website_sale/__init__.py]: > I pushed a new diff, we don't need to create with all properties the record, a copy is enough. I also changed the `'''` to `"""` as it should be in docstring (I changed the already existing ones so we don't copy paste those again, note that I introduced them myself grimacing ) I rewrote the commit message, you can see the change by comparing the PR msg and the commit one.
> 
> Do you see any issue with the changes I made?

I think the changes are fine

**Comment by rdeodoo** [comment]
@robodoo r+
